File: app/src/main/java/com/weiqi/ui/board/BoardCanvas.kt
package com.weiqi.ui.board

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextMeasurer
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.sp
import com.weiqi.engine.Board
import com.weiqi.engine.CellState
import com.weiqi.engine.Point
import com.weiqi.engine.StoneColor
import com.weiqi.ui.theme.Zen
import kotlin.math.min
import kotlin.math.roundToInt

data class BoardOverlay(
    val lastMove: Point? = null,
    val koPoint: Point? = null,
    val deadStones: Set<Point> = emptySet(),
    val territoryBlack: Set<Point> = emptySet(),
    val territoryWhite: Set<Point> = emptySet(),
    val moveNumbers: Map<Point, Int> = emptyMap(),
    /** Ghost stone shown before the user confirms a move (beginner hints mode). */
    val pending: Pair<Point, StoneColor>? = null,
    /** Markers (small filled dots) — used for tutorial diagrams. */
    val markers: Set<Point> = emptySet()
)

data class BoardAppearance(
    val board: Color = Zen.kayaWood,
    val boardEdge: Color = Zen.kayaWoodEdge,
    val ink: Color = Zen.gridInk,
    val showCoordinates: Boolean = false
) {
    companion object {
        val ClassicWood = BoardAppearance()
        val MinimalPaper = BoardAppearance(
            board = Color(0xFFFAF6EE),
            boardEdge = Color(0xFFE0DAC8),
            ink = Color(0xFF1A1A1A)
        )
        val DarkSlate = BoardAppearance(
            board = Color(0xFFD8D7D2),
            boardEdge = Color(0xFFB7B6B0),
            ink = Color(0xFF1A1A1A)
        )
        val HighContrast = BoardAppearance(
            board = Color(0xFFFFF1CD),
            boardEdge = Color(0xFFE6D496),
            ink = Color(0xFF000000)
        )
    }
}

/** A captured stone that should fade out for one frame batch. */
private data class CaptureGhost(val point: Point, val color: CellState, val key: Long)

@Composable
fun BoardCanvas(
    board: Board,
    overlay: BoardOverlay = BoardOverlay(),
    appearance: BoardAppearance = BoardAppearance.ClassicWood,
    onTap: (Point) -> Unit = {},
    modifier: Modifier = Modifier
) {
    val starPoints = remember(board.size) { starPointsFor(board.size) }
    val measurer: TextMeasurer = rememberTextMeasurer()

    // Place-in animation: last move scales from a starting size up to full.
    val placeScale = remember { Animatable(1f) }
    val lastMove = overlay.lastMove
    LaunchedEffect(lastMove, board) {
        if (lastMove != null && board.get(lastMove) != CellState.EMPTY) {
            placeScale.snapTo(0.55f)
            placeScale.animateTo(1f, animationSpec = tween(durationMillis = 180))
        } else {
            placeScale.snapTo(1f)
        }
    }

    // Capture-out animation: stones that disappeared between frames fade out.
    var prevBoard by remember { mutableStateOf(board) }
    val captureGhosts = remember { mutableStateListOf<CaptureGhost>() }
    val captureAlpha = remember { Animatable(1f) }
    LaunchedEffect(board) {
        if (prevBoard !== board && prevBoard.size == board.size) {
            val gone = ArrayList<CaptureGhost>()
            val keyBase = System.currentTimeMillis()
            for (r in 0 until board.size) for (c in 0 until board.size) {
                val before = prevBoard.get(r, c)
                val after = board.get(r, c)
                if (before != CellState.EMPTY && after == CellState.EMPTY) {
                    gone.add(CaptureGhost(Point(r, c), before, keyBase + r * board.size + c))
                }
            }
            if (gone.isNotEmpty()) {
                captureGhosts.clear()
                captureGhosts.addAll(gone)
                captureAlpha.snapTo(1f)
                captureAlpha.animateTo(0f, animationSpec = tween(durationMillis = 240))
                captureGhosts.clear()
            }
        }
        prevBoard = board
    }

    Canvas(
        modifier = modifier
            .fillMaxSize()
            .pointerInput(board.size, appearance.showCoordinates) {
                detectTapGestures { offset ->
                    val side = min(size.width, size.height).toFloat()
                    val pad = side * (if (appearance.showCoordinates) 0.085f else 0.06f)
                    val usable = side - 2 * pad
                    val step = usable / (board.size - 1)
                    val col = ((offset.x - pad) / step).roundToInt()
                    val row = ((offset.y - pad) / step).roundToInt()
                    if (row in 0 until board.size && col in 0 until board.size) {
                        onTap(Point(row, col))
                    }
                }
            }
    ) {
        val side = min(size.width, size.height)
        val pad = side * (if (appearance.showCoordinates) 0.085f else 0.06f)
        val usable = side - 2 * pad
        val step = usable / (board.size - 1)
        val stoneR = step * 0.46f
        val ink20 = appearance.ink.copy(alpha = 0.20f)
        val ink70 = appearance.ink.copy(alpha = 0.70f)

        val shadowInset = side * 0.012f
        drawSoftShadow(Offset(shadowInset, shadowInset * 1.4f),
            Size(side - 2 * shadowInset, side - 2 * shadowInset))

        drawRect(
            brush = Brush.verticalGradient(0f to appearance.board, 1f to appearance.boardEdge),
            size = Size(side, side)
        )
        drawRect(
            color = appearance.boardEdge.copy(alpha = 0.35f),
            size = Size(side, side),
            style = Stroke(width = side * 0.012f)
        )

        for (i in 0 until board.size) {
            val pos = pad + step * i
            drawLine(ink20, Offset(pad, pos), Offset(pad + usable, pos), strokeWidth = 1.2f)
            drawLine(ink20, Offset(pos, pad), Offset(pos, pad + usable), strokeWidth = 1.2f)
        }
        drawRect(
            color = ink20.copy(alpha = 0.30f),
            topLeft = Offset(pad, pad),
            size = Size(usable, usable),
            style = Stroke(width = 1.4f)
        )
        for (sp in starPoints) {
            drawCircle(ink70, radius = step * 0.085f,
                center = Offset(pad + step * sp.col, pad + step * sp.row))
        }

        if (appearance.showCoordinates) {
            val labelStyle = androidx.compose.ui.text.TextStyle(
                color = appearance.ink.copy(alpha = 0.65f),
                fontSize = (step * 0.32f).coerceAtMost(20f).sp
            )
            for (i in 0 until board.size) {
                val colLetter = colLetter(i)
                val rowLabel = (board.size - i).toString()
                val xCol = pad + step * i
                val yRow = pad + step * i
                val colLayout = measurer.measure(colLetter, labelStyle)
                val rowLayout = measurer.measure(rowLabel, labelStyle)
                drawText(colLayout, topLeft = Offset(
                    xCol - colLayout.size.width / 2f,
                    pad - colLayout.size.height - step * 0.10f
                ))
                drawText(colLayout, topLeft = Offset(
                    xCol - colLayout.size.width / 2f,
                    pad + usable + step * 0.10f
                ))
                drawText(rowLayout, topLeft = Offset(
                    pad - rowLayout.size.width - step * 0.18f,
                    yRow - rowLayout.size.height / 2f
                ))
                drawText(rowLayout, topLeft = Offset(
                    pad + usable + step * 0.18f,
                    yRow - rowLayout.size.height / 2f
                ))
            }
        }

        for (p in overlay.territoryBlack) {
            drawCircle(Color.Black.copy(alpha = 0.22f), radius = stoneR * 0.40f,
                center = Offset(pad + step * p.col, pad + step * p.row))
        }
        for (p in overlay.territoryWhite) {
            drawCircle(Color.White.copy(alpha = 0.55f), radius = stoneR * 0.40f,
                center = Offset(pad + step * p.col, pad + step * p.row))
        }

        // Diagram markers (small filled circles), e.g. tutorial liberty markers.
        for (p in overlay.markers) {
            drawCircle(
                color = Color(0xFFB23A2E).copy(alpha = 0.85f),
                radius = stoneR * 0.30f,
                center = Offset(pad + step * p.col, pad + step * p.row)
            )
        }

        // Fade-out ghosts for captured stones.
        if (captureGhosts.isNotEmpty()) {
            for (g in captureGhosts) {
                drawStone(
                    cellState = g.color,
                    center = Offset(pad + step * g.point.col, pad + step * g.point.row),
                    stoneR = stoneR,
                    ink = appearance.ink,
                    scale = 1f,
                    alpha = captureAlpha.value
                )
            }
        }

        // Stones.
        for (r in 0 until board.size) for (c in 0 until board.size) {
            val state = board.get(r, c)
            if (state == CellState.EMPTY) continue
            val center = Offset(pad + step * c, pad + step * r)
            val isLast = lastMove != null && lastMove.row == r && lastMove.col == c
            drawStone(
                cellState = state,
                center = center,
                stoneR = stoneR,
                ink = appearance.ink,
                scale = if (isLast) placeScale.value else 1f,
                alpha = 1f
            )

            if (Point(r, c) in overlay.deadStones) {
                val s = stoneR * 0.55f
                val x = Color(0xFFB23A2E)
                drawLine(x, Offset(center.x - s, center.y - s), Offset(center.x + s, center.y + s), strokeWidth = 3f)
                drawLine(x, Offset(center.x - s, center.y + s), Offset(center.x + s, center.y - s), strokeWidth = 3f)
            }
            overlay.moveNumbers[Point(r, c)]?.let { n ->
                val isBlack = state == CellState.BLACK
                val labelStyle = androidx.compose.ui.text.TextStyle(
                    color = if (isBlack) Color.White else Color.Black,
                    fontSize = (stoneR * 0.85f).coerceAtMost(18f).sp
                )
                val layout = measurer.measure(n.toString(), labelStyle)
                drawText(layout, topLeft = Offset(
                    center.x - layout.size.width / 2f,
                    center.y - layout.size.height / 2f
                ))
            }
        }

        // Ghost (preview) stone.
        overlay.pending?.let { (p, color) ->
            if (board.get(p) == CellState.EMPTY) {
                drawStone(
                    cellState = if (color == StoneColor.BLACK) CellState.BLACK else CellState.WHITE,
                    center = Offset(pad + step * p.col, pad + step * p.row),
                    stoneR = stoneR,
                    ink = appearance.ink,
                    scale = 1f,
                    alpha = 0.45f
                )
            }
        }

        overlay.lastMove?.let { p ->
            if (board.get(p) != CellState.EMPTY) {
                val state = board.get(p)
                val ringColor = if (state == CellState.BLACK) Color.White else Color.Black
                drawCircle(
                    color = ringColor,
                    radius = stoneR * 0.36f,
                    center = Offset(pad + step * p.col, pad + step * p.row),
                    style = Stroke(width = 2.2f)
                )
            }
        }
        overlay.koPoint?.let { p ->
            drawCircle(
                color = Color(0xFFB23A2E),
                radius = stoneR * 0.28f,
                center = Offset(pad + step * p.col, pad + step * p.row),
                style = Stroke(width = 2f)
            )
        }
    }
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawStone(
    cellState: CellState,
    center: Offset,
    stoneR: Float,
    ink: Color,
    scale: Float,
    alpha: Float
) {
    val r = stoneR * scale
    drawCircle(
        color = Color.Black.copy(alpha = 0.25f * alpha),
        radius = r,
        center = center.copy(x = center.x + r * 0.05f, y = center.y + r * 0.18f)
    )
    val isBlack = cellState == CellState.BLACK
    val highlight = if (isBlack) Zen.blackStoneTop else Zen.whiteStoneTop
    val body = if (isBlack) Zen.blackStoneBottom else Zen.whiteStoneBottom
    drawCircle(
        brush = Brush.radialGradient(
            0f to highlight.copy(alpha = highlight.alpha * alpha),
            1f to body.copy(alpha = body.alpha * alpha),
            center = Offset(center.x - r * 0.30f, center.y - r * 0.30f),
            radius = r * 1.4f
        ),
        radius = r,
        center = center
    )
    drawCircle(
        color = if (isBlack) Color.Black.copy(alpha = alpha) else ink.copy(alpha = 0.55f * alpha),
        radius = r,
        center = center,
        style = Stroke(width = if (isBlack) 0.6f else 1f)
    )
}

private fun androidx.compose.ui.graphics.drawscope.DrawScope.drawSoftShadow(topLeft: Offset, sz: Size) {
    val color = Color(0x1F000000)
    for (i in 0 until 3) {
        val grow = (i + 1) * 4f
        drawRect(
            color = color.copy(alpha = color.alpha / (i + 2f)),
            topLeft = Offset(topLeft.x - grow, topLeft.y - grow + 2f),
            size = Size(sz.width + grow * 2, sz.height + grow * 2)
        )
    }
}

private fun colLetter(col: Int): String {
    val skipI = 'I' - 'A'
    val ch = if (col < skipI) 'A' + col else 'A' + col + 1
    return ch.toString()
}

private fun starPointsFor(size: Int): List<Point> {
    if (size != 9 && size != 13 && size != 19) return emptyList()
    val edge = if (size == 9) 2 else 3
    val far = size - 1 - edge
    val mid = size / 2
    return when (size) {
        9 -> listOf(Point(edge, edge), Point(edge, far), Point(far, edge), Point(far, far), Point(mid, mid))
        13 -> listOf(Point(edge, edge), Point(edge, far), Point(far, edge), Point(far, far), Point(mid, mid))
        else -> listOf(
            Point(edge, edge), Point(edge, mid), Point(edge, far),
            Point(mid, edge), Point(mid, mid), Point(mid, far),
            Point(far, edge), Point(far, mid), Point(far, far)
        )
    }
}

File: lib/src/ui/board/board_canvas.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../../domain/board.dart';
import '../../domain/models.dart';
import '../theme.dart';

class BoardOverlay {
  final Point? lastMove;
  final Point? koPoint;
  final Set<Point> deadStones;
  final Set<Point> territoryBlack;
  final Set<Point> territoryWhite;
  final Map<Point, int> moveNumbers;

  /// Ghost stone shown before the user confirms a move (beginner hints mode).
  final ({Point point, StoneColor color})? pending;

  /// Markers (small filled dots) — used for tutorial diagrams.
  final Set<Point> markers;

  const BoardOverlay({
    this.lastMove,
    this.koPoint,
    this.deadStones = const {},
    this.territoryBlack = const {},
    this.territoryWhite = const {},
    this.moveNumbers = const {},
    this.pending,
    this.markers = const {},
  });
}

class BoardAppearance {
  final Color board;
  final Color boardEdge;
  final Color ink;
  final bool showCoordinates;

  const BoardAppearance({
    this.board = Zen.kayaWood,
    this.boardEdge = Zen.kayaWoodEdge,
    this.ink = Zen.gridInk,
    this.showCoordinates = false,
  });

  BoardAppearance copyWith(
          {Color? board,
          Color? boardEdge,
          Color? ink,
          bool? showCoordinates}) =>
      BoardAppearance(
        board: board ?? this.board,
        boardEdge: boardEdge ?? this.boardEdge,
        ink: ink ?? this.ink,
        showCoordinates: showCoordinates ?? this.showCoordinates,
      );

  static const classicWood = BoardAppearance();
  static const minimalPaper = BoardAppearance(
    board: Color(0xFFFAF6EE),
    boardEdge: Color(0xFFE0DAC8),
    ink: Color(0xFF1A1A1A),
  );
  static const darkSlate = BoardAppearance(
    board: Color(0xFFD8D7D2),
    boardEdge: Color(0xFFB7B6B0),
    ink: Color(0xFF1A1A1A),
  );
  static const highContrast = BoardAppearance(
    board: Color(0xFFFFF1CD),
    boardEdge: Color(0xFFE6D496),
    ink: Color(0xFF000000),
  );
}

class _CaptureGhost {
  final Point point;
  final CellState color;
  const _CaptureGhost(this.point, this.color);
}

class BoardCanvas extends StatefulWidget {
  final Board board;
  final BoardOverlay overlay;
  final BoardAppearance appearance;
  final ValueChanged<Point>? onTap;

  const BoardCanvas({
    super.key,
    required this.board,
    this.overlay = const BoardOverlay(),
    this.appearance = BoardAppearance.classicWood,
    this.onTap,
  });

  @override
  State<BoardCanvas> createState() => _BoardCanvasState();
}

class _BoardCanvasState extends State<BoardCanvas>
    with TickerProviderStateMixin {
  late final AnimationController _placeController;
  late final AnimationController _captureController;
  Board? _prevBoard;
  Point? _prevLastMove;
  List<_CaptureGhost> _captureGhosts = const [];

  @override
  void initState() {
    super.initState();
    _placeController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 180),
      value: 1,
    );
    _captureController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 240),
      value: 1,
    );
    _prevBoard = widget.board;
    _prevLastMove = widget.overlay.lastMove;
  }

  @override
  void didUpdateWidget(covariant BoardCanvas oldWidget) {
    super.didUpdateWidget(oldWidget);
    final lastMove = widget.overlay.lastMove;
    if (lastMove != _prevLastMove) {
      if (lastMove != null &&
          widget.board.cellAt(lastMove) != CellState.empty) {
        _placeController.value = 0.55;
        _placeController.animateTo(1, curve: Curves.easeOut);
      } else {
        _placeController.value = 1;
      }
      _prevLastMove = lastMove;
    }
    if (!identical(_prevBoard, widget.board) &&
        _prevBoard != null &&
        _prevBoard!.size == widget.board.size) {
      final gone = <_CaptureGhost>[];
      for (var r = 0; r < widget.board.size; r++) {
        for (var c = 0; c < widget.board.size; c++) {
          final before = _prevBoard!.cellAtRC(r, c);
          final after = widget.board.cellAtRC(r, c);
          if (before != CellState.empty && after == CellState.empty) {
            gone.add(_CaptureGhost(Point(r, c), before));
          }
        }
      }
      if (gone.isNotEmpty) {
        _captureGhosts = gone;
        _captureController.value = 1;
        _captureController.animateTo(0).then((_) {
          if (!mounted) return;
          setState(() => _captureGhosts = const []);
        });
      }
    }
    _prevBoard = widget.board;
  }

  @override
  void dispose() {
    _placeController.dispose();
    _captureController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      behavior: HitTestBehavior.opaque,
      onTapDown: (details) {
        final box = context.findRenderObject() as RenderBox?;
        if (box == null) return;
        final size = box.size;
        final side = math.min(size.width, size.height);
        final pad = side * (widget.appearance.showCoordinates ? 0.085 : 0.06);
        final usable = side - 2 * pad;
        final step = usable / (widget.board.size - 1);
        final col = ((details.localPosition.dx - pad) / step).round();
        final row = ((details.localPosition.dy - pad) / step).round();
        if (row >= 0 &&
            row < widget.board.size &&
            col >= 0 &&
            col < widget.board.size) {
          widget.onTap?.call(Point(row, col));
        }
      },
      child: AnimatedBuilder(
        animation: Listenable.merge([_placeController, _captureController]),
        builder: (context, _) => CustomPaint(
          size: Size.infinite,
          painter: _BoardPainter(
            board: widget.board,
            overlay: widget.overlay,
            appearance: widget.appearance,
            placeScale: _placeController.value,
            captureAlpha: _captureController.value,
            captureGhosts: _captureGhosts,
          ),
        ),
      ),
    );
  }
}

class _BoardPainter extends CustomPainter {
  final Board board;
  final BoardOverlay overlay;
  final BoardAppearance appearance;
  final double placeScale;
  final double captureAlpha;
  final List<_CaptureGhost> captureGhosts;

  _BoardPainter({
    required this.board,
    required this.overlay,
    required this.appearance,
    required this.placeScale,
    required this.captureAlpha,
    required this.captureGhosts,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final side = math.min(size.width, size.height);
    final pad = side * (appearance.showCoordinates ? 0.085 : 0.06);
    final usable = side - 2 * pad;
    final step = usable / (board.size - 1);
    final stoneR = step * 0.46;
    final ink20 = appearance.ink.withValues(alpha: 0.20);
    final ink70 = appearance.ink.withValues(alpha: 0.70);

    final shadowInset = side * 0.012;
    _drawSoftShadow(
      canvas,
      Offset(shadowInset, shadowInset * 1.4),
      Size(side - 2 * shadowInset, side - 2 * shadowInset),
    );

    final boardRect = Rect.fromLTWH(0, 0, side, side);
    final boardPaint = Paint()
      ..shader = LinearGradient(
        begin: Alignment.topCenter,
        end: Alignment.bottomCenter,
        colors: [appearance.board, appearance.boardEdge],
      ).createShader(boardRect);
    canvas.drawRect(boardRect, boardPaint);
    canvas.drawRect(
      boardRect,
      Paint()
        ..style = PaintingStyle.stroke
        ..color = appearance.boardEdge.withValues(alpha: 0.35)
        ..strokeWidth = side * 0.012,
    );

    final gridPaint = Paint()
      ..color = ink20
      ..strokeWidth = 1.2;
    for (var i = 0; i < board.size; i++) {
      final pos = pad + step * i;
      canvas.drawLine(Offset(pad, pos), Offset(pad + usable, pos), gridPaint);
      canvas.drawLine(Offset(pos, pad), Offset(pos, pad + usable), gridPaint);
    }
    canvas.drawRect(
      Rect.fromLTWH(pad, pad, usable, usable),
      Paint()
        ..style = PaintingStyle.stroke
        ..color = ink20.withValues(alpha: 0.30)
        ..strokeWidth = 1.4,
    );
    for (final sp in _starPointsFor(board.size)) {
      canvas.drawCircle(
        Offset(pad + step * sp.col, pad + step * sp.row),
        step * 0.085,
        Paint()..color = ink70,
      );
    }

    if (appearance.showCoordinates) {
      final color = appearance.ink.withValues(alpha: 0.65);
      final fontSize = math.min(step * 0.32, 20.0);
      for (var i = 0; i < board.size; i++) {
        final colLetter = _colLetter(i);
        final rowLabel = (board.size - i).toString();
        final xCol = pad + step * i;
        final yRow = pad + step * i;
        final colTp = _measure(colLetter, color, fontSize);
        final rowTp = _measure(rowLabel, color, fontSize);
        colTp.paint(canvas,
            Offset(xCol - colTp.width / 2, pad - colTp.height - step * 0.10));
        colTp.paint(
            canvas, Offset(xCol - colTp.width / 2, pad + usable + step * 0.10));
        rowTp.paint(canvas,
            Offset(pad - rowTp.width - step * 0.18, yRow - rowTp.height / 2));
        rowTp.paint(canvas,
            Offset(pad + usable + step * 0.18, yRow - rowTp.height / 2));
      }
    }

    for (final p in overlay.territoryBlack) {
      canvas.drawCircle(
        Offset(pad + step * p.col, pad + step * p.row),
        stoneR * 0.40,
        Paint()..color = Colors.black.withValues(alpha: 0.22),
      );
    }
    for (final p in overlay.territoryWhite) {
      canvas.drawCircle(
        Offset(pad + step * p.col, pad + step * p.row),
        stoneR * 0.40,
        Paint()..color = Colors.white.withValues(alpha: 0.55),
      );
    }

    for (final p in overlay.markers) {
      canvas.drawCircle(
        Offset(pad + step * p.col, pad + step * p.row),
        stoneR * 0.30,
        Paint()..color = const Color(0xFFB23A2E).withValues(alpha: 0.85),
      );
    }

    if (captureGhosts.isNotEmpty) {
      for (final g in captureGhosts) {
        _drawStone(
          canvas,
          cellState: g.color,
          center: Offset(pad + step * g.point.col, pad + step * g.point.row),
          stoneR: stoneR,
          ink: appearance.ink,
          scale: 1,
          alpha: captureAlpha,
        );
      }
    }

    for (var r = 0; r < board.size; r++) {
      for (var c = 0; c < board.size; c++) {
        final state = board.cellAtRC(r, c);
        if (state == CellState.empty) continue;
        final center = Offset(pad + step * c, pad + step * r);
        final isLast = overlay.lastMove != null &&
            overlay.lastMove!.row == r &&
            overlay.lastMove!.col == c;
        _drawStone(
          canvas,
          cellState: state,
          center: center,
          stoneR: stoneR,
          ink: appearance.ink,
          scale: isLast ? placeScale : 1.0,
          alpha: 1,
        );
        if (overlay.deadStones.contains(Point(r, c))) {
          final s = stoneR * 0.55;
          final paint = Paint()
            ..color = const Color(0xFFB23A2E)
            ..strokeWidth = 3;
          canvas.drawLine(Offset(center.dx - s, center.dy - s),
              Offset(center.dx + s, center.dy + s), paint);
          canvas.drawLine(Offset(center.dx - s, center.dy + s),
              Offset(center.dx + s, center.dy - s), paint);
        }
        final n = overlay.moveNumbers[Point(r, c)];
        if (n != null) {
          final isBlack = state == CellState.black;
          final fontSize = math.min(stoneR * 0.85, 18.0);
          final tp = _measure(
              n.toString(), isBlack ? Colors.white : Colors.black, fontSize);
          tp.paint(canvas,
              Offset(center.dx - tp.width / 2, center.dy - tp.height / 2));
        }
      }
    }

    final pending = overlay.pending;
    if (pending != null && board.cellAt(pending.point) == CellState.empty) {
      _drawStone(
        canvas,
        cellState: pending.color == StoneColor.black
            ? CellState.black
            : CellState.white,
        center: Offset(
            pad + step * pending.point.col, pad + step * pending.point.row),
        stoneR: stoneR,
        ink: appearance.ink,
        scale: 1,
        alpha: 0.45,
      );
    }

    final lm = overlay.lastMove;
    if (lm != null && board.cellAt(lm) != CellState.empty) {
      final state = board.cellAt(lm);
      final ringColor = state == CellState.black ? Colors.white : Colors.black;
      canvas.drawCircle(
        Offset(pad + step * lm.col, pad + step * lm.row),
        stoneR * 0.32,
        Paint()
          ..style = PaintingStyle.stroke
          ..color = ringColor
          ..strokeWidth = 1.1,
      );
    }
    final ko = overlay.koPoint;
    if (ko != null) {
      canvas.drawCircle(
        Offset(pad + step * ko.col, pad + step * ko.row),
        stoneR * 0.28,
        Paint()
          ..style = PaintingStyle.stroke
          ..color = const Color(0xFFB23A2E)
          ..strokeWidth = 2,
      );
    }
  }

  void _drawStone(
    Canvas canvas, {
    required CellState cellState,
    required Offset center,
    required double stoneR,
    required Color ink,
    required double scale,
    required double alpha,
  }) {
    final r = stoneR * scale;
    canvas.drawCircle(
      Offset(center.dx + r * 0.05, center.dy + r * 0.18),
      r,
      Paint()..color = Colors.black.withValues(alpha: 0.25 * alpha),
    );
    final isBlack = cellState == CellState.black;
    final highlight = isBlack ? Zen.blackStoneTop : Zen.whiteStoneTop;
    final body = isBlack ? Zen.blackStoneBottom : Zen.whiteStoneBottom;
    final rect = Rect.fromCircle(center: center, radius: r);
    final shader = RadialGradient(
      center: const Alignment(-0.6, -0.6),
      radius: 0.95,
      colors: [
        highlight.withValues(alpha: alpha),
        body.withValues(alpha: alpha),
      ],
    ).createShader(rect);
    canvas.drawCircle(center, r, Paint()..shader = shader);
    canvas.drawCircle(
      center,
      r,
      Paint()
        ..style = PaintingStyle.stroke
        ..color = isBlack
            ? Colors.black.withValues(alpha: 0.82 * alpha)
            : ink.withValues(alpha: 0.42 * alpha)
        ..strokeWidth = isBlack ? 0.45 : 0.7,
    );
  }

  void _drawSoftShadow(Canvas canvas, Offset topLeft, Size sz) {
    const base = Color(0x1F000000);
    for (var i = 0; i < 3; i++) {
      final grow = (i + 1) * 4.0;
      canvas.drawRect(
        Rect.fromLTWH(
          topLeft.dx - grow,
          topLeft.dy - grow + 2,
          sz.width + grow * 2,
          sz.height + grow * 2,
        ),
        Paint()..color = base.withValues(alpha: base.a / (i + 2)),
      );
    }
  }

  TextPainter _measure(String text, Color color, double fontSize) {
    final tp = TextPainter(
      text: TextSpan(
          text: text, style: TextStyle(color: color, fontSize: fontSize)),
      textDirection: TextDirection.ltr,
    );
    tp.layout();
    return tp;
  }

  String _colLetter(int col) {
    const skipI = 8; // 'I' - 'A'
    final code = col < skipI ? 0x41 + col : 0x41 + col + 1;
    return String.fromCharCode(code);
  }

  List<Point> _starPointsFor(int size) {
    if (size != 9 && size != 13 && size != 19) return const [];
    final edge = size == 9 ? 2 : 3;
    final far = size - 1 - edge;
    final mid = size ~/ 2;
    if (size == 9 || size == 13) {
      return [
        Point(edge, edge),
        Point(edge, far),
        Point(far, edge),
        Point(far, far),
        Point(mid, mid),
      ];
    }
    return [
      Point(edge, edge),
      Point(edge, mid),
      Point(edge, far),
      Point(mid, edge),
      Point(mid, mid),
      Point(mid, far),
      Point(far, edge),
      Point(far, mid),
      Point(far, far),
    ];
  }

  @override
  bool shouldRepaint(covariant _BoardPainter oldDelegate) =>
      oldDelegate.board != board ||
      oldDelegate.overlay != overlay ||
      oldDelegate.appearance != appearance ||
      oldDelegate.placeScale != placeScale ||
      oldDelegate.captureAlpha != captureAlpha ||
      oldDelegate.captureGhosts != captureGhosts;
}
